src/common/code_utils.hpp

```diff
@@ -175,4 +175,62 @@ class NonCopyable
     NonCopyable(void) = default;
 };
 
+template <typename T> class Optional
+{
+public:
+    constexpr Optional(void) = default;
+
+    Optional(T aValue)
+    {
+        new (&mStorage) T(aValue);
+        mHasValue = true;
+    }
+
+    ~Optional(void)
+    {
+        if (mHasValue)
+        {
+            GetValue().~T();
+        }
+    }
+
+    Optional(const Optional &aOther)
+    {
+        if (aOther.mHasValue)
+        {
+            new (&mStorage) T(aOther.GetValue());
+        }
+        mHasValue = aOther.mHasValue;
+    }
+
+    Optional &operator=(const Optional &aOther)
+    {
+        if (mHasValue)
+        {
+            GetValue().~T();
+        }
+        if (aOther.mHasValue)
+        {
+            new (&mStorage) T(aOther.GetValue());
+        }
+        mHasValue = aOther.mHasValue;
```

Comment:
```suggestion
        ClearValue();
        SetValue(aOther.GetValue());
```

seems these methods can be reused in multiple places.

Reply:
Done.